crates/bevy_time/src/fixed_timestep.rs

```diff
@@ -1,17 +1,109 @@
+    fn add_fixed_schedule(&mut self, timestep: FixedTimestep, schedule: impl Stage) -> &mut Self {
+        self.add_fixed_schedule_to_stage(CoreStage::Update, timestep, schedule)
+    }
+}
+
+impl TimestepAppExt for App {
+    #[track_caller]
+    fn add_fixed_schedule_to_stage<S: Stage>(
+        &mut self,
+        stage: impl StageLabel,
+        FixedTimestep { label, step }: FixedTimestep,
+        schedule: S,
+    ) -> &mut Self {
+        // If it has a label, add it to the resource so it can be modified or peeked later.
+        if let Some(label) = label {
+            let mut timesteps: Mut<FixedTimesteps> =
+                self.world.get_resource_or_insert_with(Default::default);
+            let state = FixedTimestepState {
+                step,
+                accumulator: 0.0,
+            };
+
+            // Insert the state into the map.
+            // Panic if there already was one.
+            if timesteps.insert(label, state).is_some() {
+                #[inline(never)]
+                #[track_caller]
+                fn panic(label: impl Debug) -> ! {
+                    panic!("there is already a fixed timestep labeled '{label:?}'");
+                }
+                panic(label)
+            }
+
+            let runner = move |s: &mut dyn Stage, w: &mut World| {
+                let mut state = *w.resource::<FixedTimesteps>().get(label).unwrap();
+
+                // Core looping functionality.
+                let time = w.resource::<Time>();
+                state.accumulator += time.delta_seconds_f64();
+                while state.accumulator > state.step {
+                    state.accumulator -= state.step;
+
+                    s.run(w);
+                }
+
+                // Update the resource (we've only been operating on a copy).
+                w.resource_mut::<FixedTimesteps>().insert(label, state);
+            };
+            self.add_sub_schedule(schedule.label(label).with_runner(stage, runner));
+        }
+        // If there's no label, we can keep everything local
+        // since there's no way to refer to the schedule again anyway.
+        else {
+            let mut state = FixedTimestepState {
+                step,
+                accumulator: 0.0,
+            };
+            let runner = move |sched: &mut dyn Stage, w: &mut World| {
+                // Core looping functionality.
+                let time = w.resource::<Time>();
+                state.accumulator += time.delta_seconds_f64();
+                while state.accumulator > state.step {
+                    state.accumulator -= state.step;
+
+                    sched.run(w);
+                }
+            };
+            self.add_sub_schedule(schedule.with_runner(stage, runner));
+        }
+
+        self
+    }
+}
 
 /// The internal state of each [`FixedTimestep`].
-#[derive(Debug)]
+#[derive(Debug, Clone, Copy)]
 pub struct FixedTimestepState {
     step: f64,
     accumulator: f64,
```

Comment:
Please consider switching to using `Duration` instead of floating point values for the internal state of the fixed timestep.

I know that floats are how the existing Bevy code does it, but it was a mistake. Floats are not an accurate or reliable representation of time.

My implementation in `iyes_loopless` uses `Duration`s instead, allowing it to be more exact.

It should be a very easy and noncontroversial replacement.

While we are at it, given that this PR is reworking all of FixedTimestep, I believe this small-but-important thing should be addressed.

Reply:
Good point.